libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp

```diff
@@ -0,0 +1,250 @@
+
+// handle ArmStatus
+UC_REGISTRY_BINDER(ArmStatusCb, dronecan::remoteid::ArmStatus);
+static uavcan::Subscriber<dronecan::remoteid::ArmStatus, ArmStatusCb> *arm_status_listener[HAL_MAX_CAN_PROTOCOL_DRIVERS];
+
+static void handle_arm_status(AP_UAVCAN* ap_uavcan, uint8_t node_id, const ArmStatusCb &cb);
+
+void AP_OpenDroneID::dronecan_init(AP_UAVCAN *uavcan)
+{
+    const uint8_t driver_index = uavcan->get_driver_index();
+    const uint8_t driver_mask = 1U<<driver_index;
+    if (dronecan_done_init & driver_mask) {
+        // already initialised
+        return;
+    }
+
+    dc_location[driver_index] = new uavcan::Publisher<dronecan::remoteid::Location>(*uavcan->get_node());
+    if (dc_location[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_location[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_location[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    dc_basic_id[driver_index] = new uavcan::Publisher<dronecan::remoteid::BasicID>(*uavcan->get_node());
+    if (dc_basic_id[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_basic_id[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_basic_id[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    dc_self_id[driver_index] = new uavcan::Publisher<dronecan::remoteid::SelfID>(*uavcan->get_node());
+    if (dc_self_id[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_self_id[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_self_id[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    dc_system[driver_index] = new uavcan::Publisher<dronecan::remoteid::System>(*uavcan->get_node());
+    if (dc_system[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_system[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_system[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    dc_operator_id[driver_index] = new uavcan::Publisher<dronecan::remoteid::OperatorID>(*uavcan->get_node());
+    if (dc_operator_id[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    dc_operator_id[driver_index]->setTxTimeout(uavcan::MonotonicDuration::fromMSec(20));
+    dc_operator_id[driver_index]->setPriority(uavcan::TransferPriority::OneHigherThanLowest);
+
+    arm_status_listener[driver_index] = new uavcan::Subscriber<dronecan::remoteid::ArmStatus, ArmStatusCb>(*uavcan->get_node());
+    if (arm_status_listener[driver_index] == nullptr) {
+        goto alloc_failed;
+    }
+    arm_status_listener[driver_index]->start(ArmStatusCb(uavcan, &handle_arm_status));
+
+    dronecan_done_init |= driver_mask;
+    return;
+
+alloc_failed:
+    dronecan_init_failed |= driver_mask;
+    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "OpenDroneID DroneCAN alloc failed");
```

Comment:
We should probably call `alloc_error` here.This message could easily be lost in a rush of startup messages otherwise.